wolfcrypt/src/asn.c

```diff
@@ -13743,190 +13746,99 @@ static int MatchBaseName(int type, const char* name, int nameSz,
-                            if (XMEMCMP(cur->name, base->name, base->nameSz)
-                                    == 0) {
-                                WOLFSSL_MSG("DIR alt name constraint err");
-                                return 0;
-                            }
-                            cur = cur->next;
-                        }
-                    }
-                    #endif /* !WOLFSSL_NO_ASN_STRICT */
-                    break;
-                }
-                default:
-                    break;
-            }; /* switch */
-            base = base->next;
-        }
-    }
+    if (signer->excludedNames == NULL && signer->permittedNames == NULL)
+        return 1;
 
-    /* Check against the permitted list */
-    if (signer->permittedNames != NULL) {
-        int permittedDir = 0;
-        int matchDir;
+    for (i=0; i < sizeof(nameTypes); i++) {
+        byte nameType = nameTypes[i];
+        DNS_entry* name = NULL;
+        DNS_entry  subjectDnsName;
         Base_entry* base;
-        DNS_entry* name;
 
-        /* Check each DNS name matches a permitted. */
-        name = cert->altNames;
-        while (name != NULL) {
-            int matchDns = 0;
-            int permittedDns = 0;
-            base = signer->permittedNames;
-            do {
-                /* Looking for permittedNames that are for DNS. */
-                if (base->type == ASN_DNS_TYPE) {
-                    permittedDns = 1;
-                    matchDns = MatchBaseName(ASN_DNS_TYPE,
-                                             name->name, name->len,
-                                             base->name, base->nameSz);
-                }
-                base = base->next;
-            } while (base != NULL && !matchDns);
-            /* If we found an DNS type permittedName then name must have had a
-             * match. */
-            if (permittedDns && !matchDns)
-                return 0;
-
-            if (!permittedDns)
+        switch (nameType) {
+            case ASN_DNS_TYPE:
+                /* Should it also consider CN in subject? It could use
+                 * subjectDnsName too */
+                name = cert->altNames;
                 break;
-
-            name = name->next;
-        }
-
-        /* Check each email name matches a permitted. */
-        name = cert->altEmailNames;
-        while (name != NULL) {
-            int matchEmail = 0;
-            int permittedEmail = 0;
-            base = signer->permittedNames;
-            do {
-                /* Looking for permittedNames that are for email. */
-                if (base->type == ASN_RFC822_TYPE) {
-                    permittedEmail = 1;
-                    matchEmail = MatchBaseName(ASN_DNS_TYPE,
-                                               name->name, name->len,
-                                               base->name, base->nameSz);
+            case ASN_RFC822_TYPE:
+                /* Shouldn't it validade E= in subject as well? */
```

Comment:
Current code only checks `altEmailNames`. Can you find reference to a spec that says subjectEmail should be checked too?

Reply:
I don't know about email clients but S/MIME might be similar to HTTP usage. OpenSSL does validate it:

https://github.com/openssl/openssl/blob/485d0790ac1a29a0d4e7391d804810d485890376/crypto/x509/v3_ncons.c#L280

Reply:
Legacy implementations exist where an electronic mail address is
   embedded in the subject distinguished name in an attribute of type
   emailAddress (Section 4.1.2.6).  When constraints are imposed on the
   rfc822Name name form, but the certificate does not include a subject
   alternative name, the rfc822Name constraint MUST be applied to the
   attribute of type emailAddress in the subject distinguished name.
   The ASN.1 syntax for emailAddress and the corresponding OID are
   supplied in Appendix A.